src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -68,23 +68,36 @@ public string GetReferencePath(MetadataReferenceInfo referenceInfo)
             throw new Exception($"Could not find referenced assembly {referenceInfo}");
         }
 
-        public ImmutableArray<MetadataReference> ResolveReferences(IEnumerable<MetadataReferenceInfo> references)
+        public bool TryResolveReferences(ImmutableArray<MetadataReferenceInfo> references, out ImmutableArray<MetadataReference> results)
         {
-            var referenceArray = references.ToImmutableArray();
-            CacheNames(referenceArray);
+            if (!CacheNames(references))
+            {
+                results = default;
+                return false;
+            }
 
-            var files = referenceArray.Select(r => GetReferencePath(r));
+            var builder = ImmutableArray.CreateBuilder<MetadataReference>(references.Length);
+            foreach (var reference in references)
+            {
+                var file = GetReferencePath(reference);
+                builder.Add(MetadataReference.CreateFromFile(
+                    file,
+                    new MetadataReferenceProperties(
+                        kind: MetadataImageKind.Assembly,
+                        aliases: reference.ExternAliases,
+                        embedInteropTypes: reference.EmbedInteropTypes)));
+            }
 
-            var metadataReferences = files.Select(f => MetadataReference.CreateFromFile(f)).Cast<MetadataReference>().ToImmutableArray();
-            return metadataReferences;
+            results = builder.MoveToImmutable();
+            return true;
         }
 
-        public void CacheNames(ImmutableArray<MetadataReferenceInfo> names)
+        public bool CacheNames(ImmutableArray<MetadataReferenceInfo> references)
```

Comment:
incidentally: this is pretty much exactly what PrepareTests does--go searching for a bunch of assemblies, read in their MVIDs, and collect them in a dictionary. Perf improvements are probably possible by using a similar parallelized query like what we did in there.